In [0]:
%sql
select * from main.sgfs.car_forum_valid_cleansed limit 50;

instruction input output extracted_json Answer the following question after understanding the specifics of the vehicle in question.

It is a 1997 BMW 740iL 4.4L Rear Door Locks, Gas Door. with the VIN WBAGJ8325VDL40880 and a mileage of 165,555 mi. This 1997 BMW 740iL is a 4.4L / 8 cyl / Gas engine with 5-speed Automatic Transmission (Electronic) transmission.
Delivery system: Fuel Injection Affected component: Rear Door Locks, Gas Door Customer has no operating rear door locks or gas door. The central locking system will not comm. with the assenmacher retriever. All fuses are good. I need a strategy for pin testing the central locking module in dash. Peter, Try disconnecting gas door actuator and recheck- we have seen then cause the system to act up when they fail. Greg and crew

Pete, There are fuses in the trunk, just about the battery. Check those fuses, the fuel door actuator is likely bad and caused a fuse to blow, thus shutting down functions in the GM

Peter; Try a system reset, (disconnect the battery cables and touch them together through a 10 ohm, one watt resistor for a few minutes). Reconnect the battery and see if you can communicate. If you still can't communicate, and you can communicate with everything else, the GM is bad

Have you checked fuses yet for central locking? I beleive it's the one behind glovebox or in trunk next to battery. Typically the fuel door actuator blows the fuse causing no central locking or key functions.

Did you try removing the battery cables & touching them together to reboot the system?TomThomas M. Ramies Portola Valley Garage 4170 Alpine Rd Portola Valley, CA 94028Owner/ASE certified master technician Advanced smog license L1 advanced engine performance Bimmer-Tech group Member BIMRS member IATN tech group member Rover-tech advisory group LMVIND MBZ tech member Nor-Cal Shelby Club member PCA member Porsche Technicians Group BMW CCA member Automotive Service Council Better Business Bureau member DINAN authorized BMW performance centerContact:650-851-7442 FAX 650-851-7490 tramies@pvgarage.com www.pvgarage.com pvg.mechanicnet.com From: Peter from Georgia [mailto:iATN-e00272864949el@comments.iatn.net] Sent: Wednesday, December 19, 2007 7:54 AM To: Thomas Ramies Subject: Euro-Elec: UPDATE 97 BMW 740i, Interior, rear door locks, gas door1997 BMW 740i 4.4L rear door locks, gas doorClick here to REPLY! UPDATE: All fuses tested good. Disconnecting the fuel door motor did not help. The front locks work as designed. The rear door locks are stuck in the locked position. This is when using the switch in the center console. Peter has received (4) replies to his original message so far.ORIGINAL MESSAGE: Customer has no operating rear door locks or gas door. The central locking system will not comm. with the assenmacher retriever. All fuses are good. I need a strategy for pin testing the central locking module in dash.TESTING AND RESULTS fuses:OKTECHNICAL INFORMATIONEngine:4.4 L / 8 cyl / Gas / DOHCFuel:Fuel InjectionIgnition:DistributorlessTrans:5-speed Automatic Transmission (Electronic)Mileage:165,555 miVIN:WBAGJ8325VAffected Item:rear door locks, gas door* Offer a REPLY * Read complete DETAILS * Read any REPLIES Copyright C2007 iATN. Do not reproduce/redistribute this email.This email was sent in strict accordance with our membership agreement and terms of service.

What about unlocking through the trunk with the key? Does that open just the 2 front as well or all 4 doors? These units are common for failing. It could be a faulty switch there in the middle console, but unlikely if the 2 front are working. I would start by removing the central lock fuse and see what happens, everything should open manually. I would aslo remove the switch in the middle console, these are know for getting ruined from coke, water etc.

The lock signal out of the general module (body control unit) for the rear doors and gas door is pins number 13 (black roughly 12gage wire) and pin number 14 (blue,12gauge wire)Of the

In [0]:
%sql
SELECT 
  instruction, 
  input, 
  output, 
  ai_query(
    'sg-external', -- this is an external model - Claude Sonnet 3.7
    CONCAT(
      'Using the <TASK>, <QUERY> and <RESPONSE> scrapped from an automotive forum, 
extract the meaningful gist of the content from the <RESPONSE> and construct the best response for the <QUERY>.  
This new generated response should go inside <ANSWER> </ANSWER> tags. 
Structure it in a way thats most useful to the person posting the <QUERY> just like how someone will post a response in an online forum. 
Don\'t include finishing signatures or human names.

<TASK>',
      instruction,
      '</TASK>

<QUERY>',
      input,
      '</QUERY>

<RESPONSE>',
      output,
      '</RESPONSE>'
    )
  ) AS ANSWER
FROM main.sgfs.car_forum_valid_cleansed
LIMIT 5;

instruction input output ANSWER Answer the following question after understanding the specifics of the vehicle in question.

It is a 1997 BMW 740iL 4.4L Rear Door Locks, Gas Door. with the VIN WBAGJ8325VDL40880 and a mileage of 165,555 mi. This 1997 BMW 740iL is a 4.4L / 8 cyl / Gas engine with 5-speed Automatic Transmission (Electronic) transmission.
Delivery system: Fuel Injection Affected component: Rear Door Locks, Gas Door Customer has no operating rear door locks or gas door. The central locking system will not comm. with the assenmacher retriever. All fuses are good. I need a strategy for pin testing the central locking module in dash. Peter, Try disconnecting gas door actuator and recheck- we have seen then cause the system to act up when they fail. Greg and crew

Pete, There are fuses in the trunk, just about the battery. Check those fuses, the fuel door actuator is likely bad and caused a fuse to blow, thus shutting down functions in the GM

Peter; Try a system reset, (disconnect the battery cables and touch them together through a 10 ohm, one watt resistor for a few minutes). Reconnect the battery and see if you can communicate. If you still can't communicate, and you can communicate with everything else, the GM is bad

Have you checked fuses yet for central locking? I beleive it's the one behind glovebox or in trunk next to battery. Typically the fuel door actuator blows the fuse causing no central locking or key functions.

Did you try removing the battery cables & touching them together to reboot the system?TomThomas M. Ramies Portola Valley Garage 4170 Alpine Rd Portola Valley, CA 94028Owner/ASE certified master technician Advanced smog license L1 advanced engine performance Bimmer-Tech group Member BIMRS member IATN tech group member Rover-tech advisory group LMVIND MBZ tech member Nor-Cal Shelby Club member PCA member Porsche Technicians Group BMW CCA member Automotive Service Council Better Business Bureau member DINAN authorized BMW performance centerContact:650-851-7442 FAX 650-851-7490 tramies@pvgarage.com www.pvgarage.com pvg.mechanicnet.com From: Peter from Georgia [mailto:iATN-e00272864949el@comments.iatn.net] Sent: Wednesday, December 19, 2007 7:54 AM To: Thomas Ramies Subject: Euro-Elec: UPDATE 97 BMW 740i, Interior, rear door locks, gas door1997 BMW 740i 4.4L rear door locks, gas doorClick here to REPLY! UPDATE: All fuses tested good. Disconnecting the fuel door motor did not help. The front locks work as designed. The rear door locks are stuck in the locked position. This is when using the switch in the center console. Peter has received (4) replies to his original message so far.ORIGINAL MESSAGE: Customer has no operating rear door locks or gas door. The central locking system will not comm. with the assenmacher retriever. All fuses are good. I need a strategy for pin testing the central locking module in dash.TESTING AND RESULTS fuses:OKTECHNICAL INFORMATIONEngine:4.4 L / 8 cyl / Gas / DOHCFuel:Fuel InjectionIgnition:DistributorlessTrans:5-speed Automatic Transmission (Electronic)Mileage:165,555 miVIN:WBAGJ8325VAffected Item:rear door locks, gas door* Offer a REPLY * Read complete DETAILS * Read any REPLIES Copyright C2007 iATN. Do not reproduce/redistribute this email.This email was sent in strict accordance with our membership agreement and terms of service.

What about unlocking through the trunk with the key? Does that open just the 2 front as well or all 4 doors? These units are common for failing. It could be a faulty switch there in the middle console, but unlikely if the 2 front are working. I would start by removing the central lock fuse and see what happens, everything should open manually. I would aslo remove the switch in the middle console, these are know for getting ruined from coke, water etc.

The lock signal out of the general module (body control unit) for the rear doors and gas door is pins number 13 (black roughly 12gage wire) and pin number 14 (blue,12gauge wire)Of the large c

In [0]:
%sql
create or replace view main.sgfs.forum_data_eval_vw as 
select monotonically_increasing_id() as row_id,* from main.sgfs.car_forum_valid_cleansed limit 50;


In [0]:
%sql
create or replace table main.sgfs.forum_eval_canonical_data as
SELECT 
  row_id,
  instruction, 
  input, 
  output, 
  ai_query(
    'sg-external', -- this is an external model - Claude Sonnet 3.7
    CONCAT(
      'Using the <TASK>, <QUERY> and <RESPONSE> scrapped from an automotive forum, 
extract the meaningful gist of the content from the <RESPONSE> and construct the best response for the <QUERY>.  
This new generated response should go inside <ANSWER> </ANSWER> tags. 
Structure it in a way thats most useful to the person posting the <QUERY> just like how someone will post a response in an online forum. 
Don\'t include finishing signatures or human names.

<TASK>',
      instruction,
      '</TASK>

<QUERY>',
      input,
      '</QUERY>

<RESPONSE>',
      output,
      '</RESPONSE>'
    ),
    STRUCT('temperature', 0.0, 'max_tokens', 4096)
  ) AS extracted_answer
FROM main.sgfs.forum_data_eval_vw;


num_affected_rows num_inserted_rows

In [0]:
%sql
SELECT * FROM main.sgfs.forum_eval_canonical_data;

row_id instruction input output extracted_answer 0 Answer the following question after understanding the specifics of the vehicle in question.

It is a 1997 BMW 740iL 4.4L Rear Door Locks, Gas Door. with the VIN WBAGJ8325VDL40880 and a mileage of 165,555 mi. This 1997 BMW 740iL is a 4.4L / 8 cyl / Gas engine with 5-speed Automatic Transmission (Electronic) transmission.
Delivery system: Fuel Injection Affected component: Rear Door Locks, Gas Door Customer has no operating rear door locks or gas door. The central locking system will not comm. with the assenmacher retriever. All fuses are good. I need a strategy for pin testing the central locking module in dash. Peter, Try disconnecting gas door actuator and recheck- we have seen then cause the system to act up when they fail. Greg and crew

Pete, There are fuses in the trunk, just about the battery. Check those fuses, the fuel door actuator is likely bad and caused a fuse to blow, thus shutting down functions in the GM

Peter; Try a system reset, (disconnect the battery cables and touch them together through a 10 ohm, one watt resistor for a few minutes). Reconnect the battery and see if you can communicate. If you still can't communicate, and you can communicate with everything else, the GM is bad

Have you checked fuses yet for central locking? I beleive it's the one behind glovebox or in trunk next to battery. Typically the fuel door actuator blows the fuse causing no central locking or key functions.

Did you try removing the battery cables & touching them together to reboot the system?TomThomas M. Ramies Portola Valley Garage 4170 Alpine Rd Portola Valley, CA 94028Owner/ASE certified master technician Advanced smog license L1 advanced engine performance Bimmer-Tech group Member BIMRS member IATN tech group member Rover-tech advisory group LMVIND MBZ tech member Nor-Cal Shelby Club member PCA member Porsche Technicians Group BMW CCA member Automotive Service Council Better Business Bureau member DINAN authorized BMW performance centerContact:650-851-7442 FAX 650-851-7490 tramies@pvgarage.com www.pvgarage.com pvg.mechanicnet.com From: Peter from Georgia [mailto:iATN-e00272864949el@comments.iatn.net] Sent: Wednesday, December 19, 2007 7:54 AM To: Thomas Ramies Subject: Euro-Elec: UPDATE 97 BMW 740i, Interior, rear door locks, gas door1997 BMW 740i 4.4L rear door locks, gas doorClick here to REPLY! UPDATE: All fuses tested good. Disconnecting the fuel door motor did not help. The front locks work as designed. The rear door locks are stuck in the locked position. This is when using the switch in the center console. Peter has received (4) replies to his original message so far.ORIGINAL MESSAGE: Customer has no operating rear door locks or gas door. The central locking system will not comm. with the assenmacher retriever. All fuses are good. I need a strategy for pin testing the central locking module in dash.TESTING AND RESULTS fuses:OKTECHNICAL INFORMATIONEngine:4.4 L / 8 cyl / Gas / DOHCFuel:Fuel InjectionIgnition:DistributorlessTrans:5-speed Automatic Transmission (Electronic)Mileage:165,555 miVIN:WBAGJ8325VAffected Item:rear door locks, gas door* Offer a REPLY * Read complete DETAILS * Read any REPLIES Copyright C2007 iATN. Do not reproduce/redistribute this email.This email was sent in strict accordance with our membership agreement and terms of service.

What about unlocking through the trunk with the key? Does that open just the 2 front as well or all 4 doors? These units are common for failing. It could be a faulty switch there in the middle console, but unlikely if the 2 front are working. I would start by removing the central lock fuse and see what happens, everything should open manually. I would aslo remove the switch in the middle console, these are know for getting ruined from coke, water etc.

The lock signal out of the general module (body control unit) for the rear doors and gas door is pins number 13 (black roughly 12gage wire) and pin number 14 (blue,12gauge 

In [0]:
eval_df = _sqldf.toPandas()

In [0]:
import re
def extract_answer_content(text):
    # First, try to extract content from <ANSWER> tags
    answer_match = re.search(r'<ANSWER>(.*?)</ANSWER>', text, re.DOTALL)
    if answer_match:
        return answer_match.group(1).strip()
    
    # If neither pattern matches, return None
    return None

In [0]:
eval_df['answer'] = eval_df['extracted_answer'].apply(extract_answer_content)
spark.createDataFrame(eval_df).write.mode("overwrite").saveAsTable("main.sgfs.forum_eval_canonical_data_generated")

Now that we've gathered canonical response, let's measure our models.

In [0]:
%sql
create or replace table main.sgfs.forum_eval_canonical_data_llama_base as
SELECT
  row_id,
  instruction, 
  input, 
  ai_query(
    'base_instruct_llama_sg', --base model no ft
    CONCAT(
        'Using the <TASK>, <QUERY> posted by an user in an automotive forum, 
provide the best possible response for the <QUERY>. 
Structure it in a way thats most useful to the person posting the <QUERY> just like how someone will post a response in an online forum. 
Don\'t include finishing signatures or human names.

<TASK>',
        instruction,
        '</TASK>

<QUERY>',
        input,
        '</QUERY>'
      )
    )AS answer
FROM main.sgfs.forum_data_eval_vw;


num_affected_rows num_inserted_rows

In [0]:
llama_base_eval_df = spark.sql("select * from main.sgfs.forum_eval_canonical_data_llama_base")

In [0]:
%sql
create or replace table main.sgfs.forum_eval_canonical_data_llama_ft as
SELECT 
row_id,
  instruction, 
  input, 
  ai_query(
    'sg_llama31_70B_auto',
    CONCAT(
        'Using the <TASK>, <QUERY> posted by an user in an automotive forum, 
provide the best possible response for the <QUERY>. 
Structure it in a way thats most useful to the person posting the <QUERY> just like how someone will post a response in an online forum. 
Don\'t include finishing signatures or human names.

<TASK>',
        instruction,
        '</TASK>

<QUERY>',
        input,
        '</QUERY>'
      )
    )AS answer
FROM main.sgfs.forum_data_eval_vw;


num_affected_rows num_inserted_rows

In [0]:
%sql
CREATE OR REPLACE TABLE main.sgfs.auto_judgements AS
WITH compare_data AS (
  SELECT 
    a.answer AS ft_answer, 
    b.answer AS base_answer,
    c.answer AS canon_answer, 
    a.row_id 
  FROM main.sgfs.forum_eval_canonical_data_llama_ft a 
  JOIN main.sgfs.forum_eval_canonical_data_llama_base b ON a.row_id = b.row_id
  JOIN main.sgfs.forum_eval_canonical_data_generated c ON c.row_id = a.row_id
)
SELECT 
  *,
  ai_query(
    'sg-external',
    CONCAT(
      'The following is a comparison of the generated responses from the base model and the fine-tuned model.',
      'base model answer:', base_answer, -- base model
      'fine-tuned answer:', ft_answer, -- fine tuning
      'canonical answer:', canon_answer, -- know it all
      'Your job is to compare the base model answer with the canonical answer first and produce a score called ''base_score'' based on how similar they are. Then compare the fine-tuned model answer with the canonical answer and produce a score called ''ft_score''. ',
      'The score should be a number between 0 and 100. The higher the score, the more similar they are. Similarity is based on the coverage of facts in the canonical answer. Similar facts produces a higher score.'
      'Finally produce the following JSON output: {"base_score": <base_score> , "ft_score": <ft_score>, "base_score_reasoning": <base_score_reasoning>, "ft_score_reasoning": <ft_score_reasoning>}. These are the only keys you need to produce in the JSON. Do not pack the JSON response inside a ```json``` tag. Only produce the JSON and nothing else.'
    )
  ) AS ai_result
FROM compare_data;

num_affected_rows num_inserted_rows

In [0]:
%sql
select * from main.sgfs.auto_judgements;

ft_answer base_answer canon_answer row_id ai_result Here's a step-by-step strategy for pin testing the central locking module in the dash:

1. **Check the wiring diagram**: Before starting, ensure you have a wiring diagram for your specific vehicle to identify the correct pins and wires involved in the central locking system.

2. **Locate the Central Locking Module**: Identify the central locking module in the dash. This is usually located behind the glove box or under the steering column.

3. **Access the Module**: Remove any trim or panels necessary to access the module. Be careful not to damage any wiring or components.

4. **Identify the Pins**: Using your wiring diagram, identify the pins related to the rear door locks and gas door. These are typically part of a larger connector.

5. **Test for Power**: Use a multimeter to test for power at the identified pins when the central locking system is activated. You should see voltage if the system is functioning correctly.

6. **Check Grounds**: Ensure all grounds related to the central locking system are clean and secure. A faulty ground can prevent the system from working.

7. **Test Communication**: If power is present, test the communication between the central locking module and the actuators (rear door locks and gas door). This may involve using a scan tool or monitoring voltage changes at the actuators when the locks are activated.

8. **Replace the Module**: If all tests indicate the module is faulty, replace it with a new one. Ensure it's programmed correctly for your vehicle.

9. **Reassemble**: Once repairs are made, reassemble everything in reverse order of how it was disassembled.

By following these steps, you should be able to diagnose and potentially fix the issue with your central locking system. To tackle the issue with the central locking system on your 1997 BMW 740iL, we'll need to follow a step-by-step approach to identify the problem. Since all fuses are good, we can focus on the Central Locking Module (CLM) in the dash. Here's a strategy for pin testing:

**Preparation:**

1. Consult your repair manual or a wiring diagram specific to your vehicle (VIN WBAGJ8325VDL40880) to identify the CLM location, pinout, and wiring colors.
2. Ensure you have a multimeter and a wiring diagram to help with the testing process.

**Step 1: Identify the CLM Connector**

Locate the CLM in the dash and identify the connector. It's usually a 32-pin or 40-pin connector, depending on the model year and configuration. Consult your wiring diagram to confirm the connector type and pinout.

**Step 2: Pin Testing**

Using your multimeter, perform the following tests:

1. **Power supply:** Measure the voltage at the CLM connector pins that supply power to the module (usually pins 1, 2, or 3). You should read around 12V with the ignition switch in the "ON" position.
2. **Ground:** Verify the ground connection at the CLM connector (usually pins 4, 5, or 6). You should read 0V or very close to 0V with the multimeter set to DC volts.
3. **Communication lines:** Identify the pins responsible for communication between the CLM and the Assenmacher Retriever (usually pins 10, 11, or 12). Measure the voltage on these pins with the ignition switch in the "ON" position. You should read a voltage signal (around 5V or 12V) indicating communication activity.
4. **Output signals:** Measure the voltage on the pins that control the rear door locks and gas door (usually pins 20, 21, or 22). You should read a voltage signal (around 12V) when the central locking system is activated.

**Step 3: Analyze Results**

Based on your test results, you can start to identify the problem:

* If you don't read any voltage on the power supply pins, check the wiring and fuses again.
* If the ground connection is faulty, repair or replace the ground wire.
* If the communication lines show no activity, check the wiring and connections between the CLM and the Assenmacher Retriever.
* If the output signals are not present,

In [0]:
%sql
create or replace view main.sgfs.auto_judgements_vw as
select 
get_json_object(ai_result,'$.base_score') as base_score,
get_json_object(ai_result,'$.ft_score') as ft_score,
get_json_object(ai_result,'$.base_score_reasoning') as base_score_reasoning,
get_json_object(ai_result,'$.ft_score_reasoning') as ft_score_reasoning
from main.sgfs.auto_judgements;

In [0]:
%sql
select 
avg(base_score) as avg_base_score,
avg(ft_score) as avg_ft_score,
median(base_score) as median_base_score,
median(ft_score) as median_ft_score,
round((avg(ft_score) - avg(base_score))/avg(base_score) * 100) as avg_improvement_percent,
round((median(ft_score) - median(base_score))/median(base_score) * 100) as median_improvement_percent
from main.sgfs.auto_judgements_vw;

avg_base_score avg_ft_score median_base_score median_ft_score avg_improvement_percent median_improvement_percent 47.8 54.4 45.0 60.0 14.0 33.0

The following items are notably important.
1. Using distillation, we can improve the performance of the base model on our specific task by a range of ~15-33%.
2. This jump is **incredibly high** considering 33%  is the **median improvement** and that we've favorably improved the model responses in *atleast* 1 in 3 questions!
3. We also used the [most stringent model](https://www.reddit.com/r/LocalLLaMA/comments/1j1npv1/llms_grading_other_llms/) as a judge. 


<img src=https://i.redd.it/yyy9616149me1.jpeg>
